arviz/sel_utils.py

```diff
@@ -0,0 +1,224 @@
+    else:
+        sel = ""
+        base = "{}{}"
+    return base.format(var_name, sel)
+
+
+def purge_duplicates(list_in):
+    """Remove duplicates from list while preserving order.
+
+    Parameters
+    ----------
+    list_in: Iterable
+
+    Returns
+    -------
+    list
+        List of first occurrences in order
+    """
+    # Algorithm taken from Stack Overflow,
+    # https://stackoverflow.com/questions/480214. Content by Georgy
+    # Skorobogatov (https://stackoverflow.com/users/7851470/georgy) and
+    # Markus Jarderot
+    # (https://stackoverflow.com/users/22364/markus-jarderot), licensed
+    # under CC-BY-SA 4.0.
+    # https://creativecommons.org/licenses/by-sa/4.0/.
+
+    seen = set()
+    seen_add = seen.add
+    return [x for x in list_in if not (x in seen or seen_add(x))]
+
+
+def _dims(data, var_name, skip_dims):
+    return [dim for dim in data[var_name].dims if dim not in skip_dims]
+
+
+def _zip_dims(new_dims, vals):
+    return [{k: v for k, v in zip(new_dims, prod)} for prod in product(*vals)]
+
+
+def xarray_sel_iter(data, var_names=None, combined=False, skip_dims=None, reverse_selections=False):
+    """Convert xarray data to an iterator over variable names and selections.
+
+    Iterates over each var_name and all of its coordinates, returning the variable
+    names and selections that allow properly obtain the data from ``data`` as desired.
+
+    Parameters
+    ----------
+    data : xarray.Dataset
+        Posterior data in an xarray
+
+    var_names : iterator of strings (optional)
+        Should be a subset of data.data_vars. Defaults to all of them.
+
+    combined : bool
+        Whether to combine chains or leave them separate
+
+    skip_dims : set
+        dimensions to not iterate over
+
+    reverse_selections : bool
+        Whether to reverse selections before iterating.
+
+    Returns
+    -------
+    Iterator of (var_name: str, selection: dict(str, any))
+        The string is the variable name, the dictionary are coordinate names to values,.
+        To get the values of the variable at these coordinates, do
+        ``data[var_name].sel(**selection)``.
+    """
+    if skip_dims is None:
+        skip_dims = set()
+
+    if combined:
+        skip_dims = skip_dims.union({"chain", "draw"})
+    else:
+        skip_dims.add("draw")
+
+    if var_names is None:
+        if isinstance(data, xr.Dataset):
+            var_names = list(data.data_vars)
+        elif isinstance(data, xr.DataArray):
+            var_names = [data.name]
+            data = {data.name: data}
+
+    for var_name in var_names:
+        if var_name in data:
+            new_dims = _dims(data, var_name, skip_dims)
+            vals = [purge_duplicates(data[var_name][dim].values) for dim in new_dims]
+            dims = _zip_dims(new_dims, vals)
+            if reverse_selections:
+                dims = reversed(dims)
+
+            for selection in dims:
+                yield var_name, selection
```

Comment:
this should somehow return also the indices (and respect the index origin)

Reply:
I agree on returning indices, I would not enforce index_origin here (the coordinate label would respect index_origin if integer). I personally don't like having indices that do not correspond to the ones used to index/slice and therefore, as I said above, I would use regular indices and eventually (and optionally) add `index_origin`

Reply:
That sounds more robust way to handle this